astropy/stats/bayesian_blocks.py

```diff
@@ -0,0 +1,505 @@
+        N = len(t)
+        best = np.zeros(N, dtype=float)
+        last = np.zeros(N, dtype=int)
+
+        #-----------------------------------------------------------------
+        # Start with first data cell; add one cell at each iteration
+        #-----------------------------------------------------------------
+        for R in range(N):
+            # Compute fit_vec : fitness of putative last block (end at R)
+            kwds = {}
+
+            # T_k: width/duration of each block
+            if 'T_k' in self._fitness_args:
+                kwds['T_k'] = block_length[:R + 1] - block_length[R + 1]
+
+            # N_k: number of elements in each block
+            if 'N_k' in self._fitness_args:
+                kwds['N_k'] = np.cumsum(x[:R + 1][::-1])[::-1]
+
+            # a_k: eq. 31
+            if 'a_k' in self._fitness_args:
+                kwds['a_k'] = 0.5 * np.cumsum(ak_raw[:R + 1][::-1])[::-1]
+
+            # b_k: eq. 32
+            if 'b_k' in self._fitness_args:
+                kwds['b_k'] = - np.cumsum(bk_raw[:R + 1][::-1])[::-1]
+
+            # c_k: eq. 33
+            if 'c_k' in self._fitness_args:
+                kwds['c_k'] = 0.5 * np.cumsum(ck_raw[:R + 1][::-1])[::-1]
+
+            # evaluate fitness function
+            fit_vec = self.fitness(**kwds)
+
+            A_R = fit_vec - self.prior(R + 1, N)
+            A_R[1:] += best[:R]
+
+            i_max = np.argmax(A_R)
+            last[R] = i_max
+            best[R] = A_R[i_max]
+
+        #-----------------------------------------------------------------
+        # Now find changepoints by iteratively peeling off the last block
+        #-----------------------------------------------------------------
+        change_points = np.zeros(N, dtype=int)
+        i_cp = N
+        ind = N
+        while True:
+            i_cp -= 1
+            change_points[i_cp] = ind
+            if ind == 0:
+                break
+            ind = last[ind - 1]
+        change_points = change_points[i_cp:]
+
+        return edges[change_points]
+
+
+class Events(FitnessFunc):
+    """Bayesian blocks fitness for binned or unbinned events
+
+    Parameters
+    ----------
+    p0 : float (optional)
+        False alarm probability, used to compute the prior on N
+        (see eq. 21 of Scargle 2012).  Default prior is for p0 = 0.
+    gamma : float (optional)
+        If specified, then use this gamma to compute the general prior form,
+        p ~ gamma^N.  If gamma is specified, p0 is ignored.
+    """
+    def fitness(self, N_k, T_k):
+        # eq. 19 from Scargle 2012
+        return N_k * (np.log(N_k) - np.log(T_k))
+
+    def validate_input(self, t, x, sigma):
+        t, x, sigma = super(Events, self).validate_input(t, x, sigma)
+        if x is not None and np.any(x % 1 > 0):
+            raise ValueError("x must be integer counts for fitness='events'")
+        return t, x, sigma
+
+    def p0_prior(self, N, Ntot):
+        # eq. 21 from Scargle 2012
```

Comment:
This was addressed in this astroML pull request: https://github.com/astroML/astroML/pull/20
The ApJ paper has an error.